src/Microsoft.AspNetCore.Routing/EndpointConstraints/EndpointConstraintEndpointSelector.cs

```diff
@@ -72,31 +82,60 @@ public Endpoint SelectBestCandidate(HttpContext context, IReadOnlyList<Endpoint>
             }
         }
 
-        private IReadOnlyList<Endpoint> EvaluateEndpointConstraints(
+        private IReadOnlyList<EndpointSelectorCandidate> EvaluateEndpointConstraints(
             HttpContext context,
-            IReadOnlyList<Endpoint> endpoints)
+            CandidateSet candidateSet)
         {
             var candidates = new List<EndpointSelectorCandidate>();
 
             // Perf: Avoid allocations
-            for (var i = 0; i < endpoints.Count; i++)
+            for (var i = 0; i < candidateSet.Count; i++)
             {
-                var endpoint = endpoints[i];
-                var constraints = _endpointConstraintCache.GetEndpointConstraints(context, endpoint);
-                candidates.Add(new EndpointSelectorCandidate(endpoint, constraints));
+                if (candidateSet[i].IsValidCandiate)
+                {
+                    var endpoint = candidateSet[i].Endpoint;
+                    var constraints = _endpointConstraintCache.GetEndpointConstraints(context, endpoint);
+                    candidates.Add(new EndpointSelectorCandidate(
+                        endpoint,
+                        candidateSet[i].Score,
+                        candidateSet[i].Values,
+                        constraints));
+                }
             }
 
             var matches = EvaluateEndpointConstraintsCore(context, candidates, startingOrder: null);
 
-            List<Endpoint> results = null;
+            List<EndpointSelectorCandidate> results = null;
             if (matches != null)
             {
-                results = new List<Endpoint>(matches.Count);
-                // Perf: Avoid allocations
-                for (var i = 0; i < matches.Count; i++)
+                results = new List<EndpointSelectorCandidate>(matches.Count);
+
+                // We need to disambiguate based on 'score' - take the first value of 'score'
+                // and then we only copy matches while they have the same score. This accounts
+                // for a difference in behavior between new routing and old.
```

Comment:
I replaced groups with a somewhat friendlier concept. see the comments on `Candidate.Score`.